azure-discovery/src/main/java/io/crate/azure/discovery/AzureUnicastHostsProvider.java

```diff
@@ -0,0 +1,282 @@
+                    return hostType;
+                }
+            }
+            return null;
+        }
+    }
+
+    private static final TimeValue DEFAULT_REFRESH_TIME = TimeValue.timeValueSeconds(5L);
+    private static final HostType DEFAULT_HOST_TYPE = HostType.PRIVATE_IP;
+    private static final String DEFAULT_DISCOVERY_METHOD = AzureDiscovery.VNET;
+
+    private final AzureComputeService azureComputeService;
+    private final TransportService transportService;
+    private final NetworkService networkService;
+    private final Version version;
+    private final TimeValue refreshInterval;
+
+    private DiscoNodeCache cache;
+
+    private final String resourceGroup;
+    private final HostType hostType;
+    private final String discoveryMethod;
+
+    @Inject
+    public AzureUnicastHostsProvider(Settings settings,
+                                     AzureComputeService azureComputeService,
+                                     TransportService transportService,
+                                     NetworkService networkService,
+                                     Version version) {
+        super(settings);
+        this.azureComputeService = azureComputeService;
+        this.transportService = transportService;
+        this.networkService = networkService;
+        this.version = version;
+
+        refreshInterval = settings.getAsTime(Discovery.REFRESH, DEFAULT_REFRESH_TIME);
+        resourceGroup = settings.get(AzureComputeService.Management.RESOURCE_GROUP_NAME);
+        hostType = HostType.fromString(settings.get(Discovery.HOST_TYPE, DEFAULT_HOST_TYPE.name()));
+        discoveryMethod = settings.get(Discovery.DISCOVERY_METHOD, DEFAULT_DISCOVERY_METHOD);
+    }
+
+    /**
+     * We build the list of Nodes from Azure Management API
+     * List of discovery nodes is cached.
+     * The cache time can be controlled using `cloud.azure.refresh_interval` setting.
+     */
+    @Override
+    public List<DiscoveryNode> buildDynamicNodes() {
+        if (cache == null) {
+            cache = new DiscoNodeCache(refreshInterval, Collections.<DiscoveryNode>emptyList());
+            cache.refresh();
+        }
+        return cache.getOrRefresh();
+    }
+
+    private class DiscoNodeCache extends SingleObjectCache<List<DiscoveryNode>> {
+
+        protected DiscoNodeCache(TimeValue refreshInterval, List<DiscoveryNode> initialValue) {
+            super(refreshInterval, initialValue);
+        }
+
+        @Override
+        protected List<DiscoveryNode> refresh() {
+            ArrayList<DiscoveryNode> nodes = new ArrayList<>();
+            InetAddress ipAddress = null;
+            try {
+                ipAddress = networkService.resolvePublishHostAddresses(null);
+                logger.trace("ip of current node: [{}]", ipAddress);
+            } catch (IOException e) {
+                // We can't find the publish host address... Hmmm. Too bad :-(
+                logger.trace("exception while finding ip", e);
```

Comment:
in this case there will be probably an NPE below. should probably return empty here?
also i think this shuld be at least WARN level